lib/recurly/schemas.js

```diff
@@ -0,0 +1,156 @@
+const inspect = require('util').inspect
+const utils = require('./utils')
+
+/**
+ * Represents a primitive:
+ *   - String
+ *   - Number
+ *   - Boolean
+ *   - Object (plain js object)
+ *   - Array (of primitives only)
+ */
+class PrimitiveProperty {
+  isValid (value) {
+    return true
+  }
+
+  cast (value) {
+    return value
+  }
+}
+
+/**
+ * Represents a Date in the schema
+ */
+class DateProperty {
+  isValid (value) {
+    return typeof value === 'string'
+  }
+  cast (value) {
+    return new Date(value)
+  }
+}
+
+/**
+ * Represents an embedded Resource in the schema
+ */
+class ResourceProperty {
+  constructor (type) {
+    this.type = type
+  }
+
+  isValid (value) {
+    if (Array.isArray(value)) return false
+    return typeof value === 'object'
+  }
+
+  cast (value) {
+    return this.type.cast(value)
+  }
+}
+
+/**
+ * Represents an array of other schema properties.
+ *
+ * @param {Property} type - Should be Property for element type
+ */
+class ArrayProperty {
+  constructor (type) {
+    this.type = type
+  }
+
+  isValid (value) {
+    if (!Array.isArray(value)) return false
+    if (value.length === 0) return true
+    return this.type.isValid(value[0])
```

Comment:
Why not loop through the array to check if all the items are valid?

Reply:
It's kind of a tradeoff. I feel like the purpose of this check is to be a quick way to see if the OpenAPI spec is correctly returning what it's describing. I feel like a situation in which the API would be returning arrays of different types on accident would be the result of a bug that should be caught by something in the server's tests. It just speaks to a bigger problem. It just seems like a lot of extra checks when 99.9999% of the time they are gonna be the same. I could go either way, but I think this satisfies catching the types of mistakes i'm looking for.

Reply:
Yup makes perfect sense. I wasn't sure if this was something that would be caught on the server side or not.